### PR TITLE
Add unit tests

### PR DESCRIPTION
**Pull request by rosen-vladimirov**
- Add projectVersion autocompletion tests.
- Add WPSdk tests
- Add unzip tests - overwriteExistingFiles option and fix issue with usage
- Add unzip tests - caseSensitive option and fix usage
- Add init command tests - validate all config and mandatory files exists after init

### CODE REVIEW

File: test/file-system.ts [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+import util = require("util");
+import hostInfo = require("../lib/common/host-info");
+var assert = require("chai").assert;
+
+var fileSystemFile = require("../lib/common/file-system");
+var cordovaBlankTemplateZipFile = path.join(__dirname, "../resources/ProjectTemplates/Telerik.Mobile.Cordova.Blank.zip");
+var cordovaBlankTemplateZipFileIncorrectName = path.join(__dirname, "../resources/ProjectTemplates/Telerik.Mobile.Cordova.blank.zip");
+import childProcessLib = require("../lib/common/child-process");
+import staticConfigLib = require("../lib/config");
+var isOsCaseInsensitive = hostInfo.isLinux();
```
Comment (teobugslayer) [inline test/file-system.ts]: This code is correct, but for the lulz: https://msdn.microsoft.com/en-us/library/windows/desktop/aa363858(v=vs.85).aspx and scroll down to FILE_FLAG_POSIX_SEMANTICS

File: test/project-properties-service.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
+///<reference path=".d.ts"/>
+
```
Comment (teobugslayer) [inline test/project-properties-service.ts]: "Use strict"; for the win!

Comment (rosen-vladimirov) [inline test/project-properties-service.ts]: :cat:  nice catch

File: test/project.ts [hunk 3]
```diff
@@ -54,13 +72,35 @@ function createTestInjector(): IInjector {
...
+		},
+
+		getSupportedFrameworks: (): IFuture<Server.FrameworkVersion[]> => {
+			return (() => {
+				return [{ DisplayName: "version_3_2_0", Version: "3.2.0" }, { DisplayName: "version_3_5_0", Version: "3.5.0" }, { DisplayName: "version_3_7_0", Version: "3.7.0" }]
+			}).future<Server.FrameworkVersion[]>()();
+		},
+
+		getDisplayNameForVersion: (version: string): IFuture<string> => {
+			return ((): string => {
```
Comment (teobugslayer) [inline test/project.ts]: future.fromResult for shorter code?

File: test/project.ts [hunk 4]
```diff
@@ -150,7 +191,137 @@ describe("project integration tests", () => {
...
+			project.updateProjectPropertyAndSave("set", "FrameworkVersion", ["3.5.0"]).wait();
+			
+			assert.strictEqual(project.projectData.WPSdk, "8.0", "WPSdk must be downgraded to 8.0 when downgrading Cordova version from 3.7.0");
+			assert.strictEqual(project.projectData.FrameworkVersion, "3.5.0", "Cordova version should have been migrated to 3.7.0");
+		});
+
+		it("updates FrameworkVersion to 3.7.0 when WPSdk is updated to 8.1",() => {
+			var prompter = testInjector.resolve("prompter");
+			prompter.confirmResult = true;
+			var options: any = require("./../lib/common/options");
```
Comment (teobugslayer) [inline test/project.ts]: ./ :(

File: test/project.ts [hunk 5]
```diff
@@ -150,7 +191,137 @@ describe("project integration tests", () => {
...
+
+			project.updateProjectPropertyAndSave("set", "WPSdk", ["8.1"]).wait();
+			assert.strictEqual(project.projectData.FrameworkVersion, "3.7.0", "Cordova version must be updated to 3.7.0 when WPSdk is updated to 8.1");
+			assert.strictEqual(project.projectData.WPSdk, "8.1", "WPSdk should have been updated to 8.1");
+		});
+
+		it("does not update FrameworkVersion to 3.7.0 when trying to update WPSdk to 8.1 but user refuses",() => {
+			var prompter = testInjector.resolve("prompter");
+			prompter.confirmResult = false;
+			var options: any = require("./../lib/common/options");
```
Comment (teobugslayer) [inline test/project.ts]: ./ :(

File: test/project.ts [hunk 6]
```diff
@@ -150,7 +191,137 @@ describe("project integration tests", () => {
...
+			project.updateProjectPropertyAndSave("set", "FrameworkVersion", ["3.5.0"]).wait();
+			assert.isFalse(prompter.confirmCalled, "We have prompted for confirmation to change WPSdk to 8.0, but it is already 8.0. DO NOT PROMPT HERE!");
+			assert.strictEqual(project.projectData.FrameworkVersion, "3.5.0", "Cordova version should have been migrated to 3.5.0");
+		});
+
+		it("does not prompt for Cordova upgrade to 3.7.0 when WPSdk is upgraded to 8.1 and FrameworkVersion is already 3.7.0",() => {
+			var prompter = testInjector.resolve("prompter");
+			prompter.confirmCalled = false;
+			prompter.confirmResult = true;
+			var options: any = require("./../lib/common/options");
```
Comment (teobugslayer) [inline test/project.ts]: ./ - :scream_cat:

File: test/project.ts [hunk 7]
```diff
@@ -187,9 +358,78 @@ describe("project integration tests", () => {
 				var tnsDir = path.join(projectDir, "tns_modules");
 				assert.isTrue(fs.existsSync(tnsDir), "NativeScript TypeScript.Blank template does not contain mandatory 'tns_modules' directory. This directory is required in init command. You should check if this is problem with the template or change init command to use another file.");
 			});
+
+			it("existing TypeScript.Blank template has project files after init",() => {
+				var options: any = require("./../lib/common/options");
```
Comment (teobugslayer) [inline test/project.ts]: :cry:

File: test/project.ts [hunk 8]
```diff
@@ -187,9 +358,78 @@ describe("project integration tests", () => {
...
+				options.path = tempFolder;
+				options.template = "TypeScript.Blank";
+				options.appid = "com.telerik.Test";
+				var project: Project.IProject = testInjector.resolve("project");
+				project.createNewProject(projectName, projectConstants.TARGET_FRAMEWORK_IDENTIFIERS.NativeScript).wait();
+				var projectDir = project.getProjectDir().wait();
+				var projectFile = path.join(projectDir, ".abproject");
+				var abignoreFile = path.join(projectDir, ".abignore");
+				fs.unlinkSync(projectFile);
+				fs.unlinkSync(abignoreFile);
```
Comment (teobugslayer) [inline test/project.ts]: what about the .debug.* and .release.* files?

File: test/project.ts [hunk 9]
```diff
@@ -187,9 +358,78 @@ describe("project integration tests", () => {
...
+				options.path = projectDir;
+				project.initializeProjectFromExistingFiles(projectConstants.TARGET_FRAMEWORK_IDENTIFIERS.NativeScript).wait();
+				assert.isTrue(fs.existsSync(projectFile), "After initialization, project does not have .abproject file.");
+				assert.isTrue(fs.existsSync(abignoreFile), "After initialization, project does not have .abignore file.");
+			});
+
+			it("existing project has .abproject and .abignore files after init",() => {
+				var options: any = require("./../lib/common/options");
```
Comment (teobugslayer) [inline test/project.ts]: :sigh:

Comment (teobugslayer) [inline test/project.ts]: Consider adding the repeating test bootstraping code in before() callback

File: test/project.ts [hunk 10]
```diff
@@ -327,13 +567,13 @@ function getProjectData(): IProjectData {
 		"AndroidVersionCode": "1",
 		"iOSDeviceFamily": ["1", "2"],
 		"iOSBackgroundMode": [],
-		"ProjectTypeGuids": "{F0A65104-D4F4-4012-B799-F612D75820F6}",
-		"FrameworkVersion": "0.4.0",
+		"ProjectTypeGuids": "{070BCB52-5A75-4F8C-A973-144AF0EAFCC9}",
```
Comment (teobugslayer) [inline test/project.ts]: I am terrified this actually "worked".

File: test/project.ts (hunk 8 above)
Comment (rosen-vladimirov) [inline test/project.ts]: they are not part of nativescript templates (I don't know if this is correct)

**Comment by teobugslayer** [comment]
Good, good! :+1:

**Pull request merged**